src/onyx/datomic/client.clj

```diff
@@ -0,0 +1,101 @@
+    (d/client  {:server-type :cloud
+                :region region
+                :system system
+                :query-group (or query-group (System/getenv "DATOMIC_CLOUD_QUERY_GROUP") system)
+                :endpoint (or endpoint (format "http://entry.%s.%s.datomic.net:8182/" system region))
+                :proxy-port (or proxy-port 8182)})))
+
+(defn- _peer-server-client [{:keys [datomic-client/access-key
+                                    datomic-client/secret
+                                    datomic-client/endpoint] :as datomic-config}]
+  (let [access-key (or access-key (System/getenv "DATOMIC_CLIENT_ACCESS_KEY"))
+        secret (or secret (System/getenv "DATOMIC_CLIENT_SECRET"))
+        endpoint (or endpoint (System/getenv "DATOMIC_CLIENT_ENDPOINT"))]
+    (d/client  {:server-type :peer-server
+                :access-key access-key
+                :secret secret
+                :endpoint endpoint})))
+
+(def cloud-client (memoize _cloud-client))
+(def peer-server-client (memoize _peer-server-client))
+
+(defn- safe-connect-cloud [{:keys [datomic-cloud/db-name] :as datomic-config}]
+  (when (nil? db-name)
+    (throw (ex-info "either :datomic-cloud/db-name is required to connect." datomic-config)))
+  (d/connect (cloud-client datomic-config) {:db-name db-name}))
+
+(defn- safe-connect-client [{:keys [datomic-client/db-name] :as datomic-config}]
+  (when (nil? db-name)
+    (throw (ex-info ":datomic-client/db-name is required to connect." datomic-config)))
+  (log/info "Connecting to database " db-name)
+  (d/connect (peer-server-client datomic-config) {:db-name db-name}))
+
+(def not-implemented-yet #(throw (ex-info "not implmented yet" {})))
+
+(defrecord DatomicClient [lib-type]
+  dp/DatomicHelpers
+  (cas-key [_] :db/cas)
+  (create-database [_ datomic-config]
+    (if (= :cloud lib-type)
+      (d/create-database (cloud-client datomic-config) {:db-name (:datomic-cloud/db-name datomic-config)})
+      (throw (ex-info "Datomic client for On-Prem doesn't support create-database. Use peer API." datomic-config))))
+  (delete-database [_ datomic-config]
+    (if (= :cloud lib-type)
+      (d/delete-database (cloud-client datomic-config) {:db-name (:datomic-cloud/db-name datomic-config)})
+      (throw (ex-info "Datomic client for On-Prem doesn't support delete-database. Use peer API." datomic-config))))
+  (instance-of-datomic-function? [this v] false)
+  (next-t [_ db]
+    (:next-t db))
+  (safe-connect [_ datomic-config]
+    (if (= :cloud lib-type)
+      (safe-connect-cloud datomic-config)
+      (safe-connect-client datomic-config)))
+  (safe-as-of [_ datomic-config conn]
+    (if-let [t (:datomic/t datomic-config)]
+      (d/as-of (d/db conn) t)
+      (throw (ex-info ":datomic/t missing from write-datoms datomic-config." datomic-config))))
+  (transact [this conn data]
+    (d/transact conn {:tx-data data}))
+  (transact-async [this conn data]
+    (future (d/transact conn {:tx-data data})))
```

Comment:
Is `transact-async` not supported with this version? (hence the future?)

Reply:
Client API has sync and async version. async version of `transact` returns a channel. Peer API's  `transact` returns the result wrapped by future, but Client API sync version's returns actual result. So I decided to wrap it with future, and took out `deref` from the caller of `datomic.client.api/transact`.